pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -0,0 +1,239 @@
+		return err
+	}
+	return w.traversePluginDir()
+}
+
+func (w *Watcher) registerPlugin(socketPath string) error {
+	//TODO: Implement rate limiting to mitigate any DOS kind of attacks.
+	glog.V(2).Infof("registerPlugin called for socketPath: %s", socketPath)
+	client, conn, err := dial(socketPath)
+	if err != nil {
+		glog.Warningf("Dial failed at socket %s, err: %v", socketPath, err)
+		return err
+	}
+	defer conn.Close()
+	stream, err := client.GetInfo(context.Background())
+	if err != nil {
+		glog.Errorf("Failed to start GetInfo grpc stream at socket %s, err: %v", socketPath, err)
+		return err
+	}
+	defer stream.CloseSend()
+	rqst, err := stream.Recv()
+	if err != nil {
+		glog.Errorf("Failed to receive stream at socket %s, err: %v", socketPath, err)
+		return err
+	}
+	resp := new(watcherapi.RegistrationStatus)
+	if handlerCbkFn, ok := w.handlers[rqst.Type]; ok {
+		var versions []string
+		for _, version := range rqst.SupportedVersions {
+			versions = append(versions, version)
+		}
+		// calls handler callback to verify registration request
+		if verifyErr := handlerCbkFn(rqst.Name, versions, socketPath); verifyErr != nil {
+			resp.PluginRegistered = false
+			resp.Error = fmt.Sprintf("Plugin registration failed with err: %v", verifyErr)
+		} else {
+			resp.PluginRegistered = true
+		}
+	} else {
+		resp.PluginRegistered = false
+		resp.Error = fmt.Sprintf("No handler found registered for plugin type: %s, socket: %s", rqst.Type, socketPath)
+	}
+	glog.V(2).Infof("send register resp: %v", resp)
+	if err := stream.Send(resp); err != nil {
+		glog.Errorf("Failed to send registration status at socket %s, err: %v", socketPath, err)
+		return err
+	}
+
+	for {
+		_, err := stream.Recv()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			glog.Errorf("Recv failed. err: %v", err)
+			return err
+		}
+	}
+
+	if !resp.PluginRegistered {
+		glog.Errorf("Registration failed for plugin type: %s, name: %s, socket: %s, err: %s", rqst.Type, rqst.Name, socketPath, resp.Error)
+		return fmt.Errorf("%s", resp.Error)
+	}
+	glog.Infof("Successfully registered plugin for plugin type: %s, name: %s, socket: %s", rqst.Type, rqst.Name, socketPath)
+	return nil
+}
+
+// Start watches for the creation of plugin sockets at the path
+func (w *Watcher) Start() error {
+	glog.V(2).Infof("Plugin Watcher Start at %s", w.path)
+	w.stopCh = make(chan interface{})
+
+	// Creating the directory to be watched if it doesn't exist yet,
+	// and walks through the directory to discover the existing plugins.
+	if err := w.init(); err != nil {
+		return err
+	}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("failed to start plugin watcher, err: %v", err)
+	}
+
+	err = watcher.Add(w.path)
+	if err != nil {
+		watcher.Close()
+		return fmt.Errorf("failed to start plugin watcher, err: %v", err)
+	}
+	w.watcher = watcher
+	startStatusCh := make(chan interface{})
+	go func(watcher *fsnotify.Watcher) {
+		close(startStatusCh)
```

Comment:
Any reason for using this pattern? I don't see this very much and I'm not sure what you are protecting against

Reply:
To ensure that goroutine has got fired before leaving this start method. This helped get away with the infinite loop in the unit test that you were commenting about earlier.

Reply:
Can you explain a bit more the context of what triggers that infinite loop?
I don't understand why that pattern would protect against an infinite loop

Reply:
?

Reply:
nvm, forget that infinite loop. This is to ensure that goroutine has got fired up before returning from this function. And this is being ensured to make sure that if `Stop` is invoked only after goroutine has got triggered. HTH!

Reply:
I don't understand what race is there between the goroutine start and calling the `Stop` function

Reply:
If Start and Stop are invoked in succession, it may happen that Goroutine has not actually started executing but Stop has got completed.

Reply:
In stop you are waiting for `w.stopCh` to be closed how can `Stop` complete if that goroutine does not start?

Reply:
even before that `close(w.stopCh)` will fail in such a case.